californium-core/src/main/java/org/eclipse/californium/core/CoapClient.java

```diff
@@ -846,7 +849,15 @@ private Request ifNoneMatch(Request request) {
 	private CoapObserveRelation observeAndWait(Request request, CoapHandler handler) {
 		Endpoint outEndpoint = getEffectiveEndpoint(request);
 		CoapObserveRelation relation = new CoapObserveRelation(request, outEndpoint);
-		request.addMessageObserver(new ObserveMessageObserverImpl(handler, relation));
+		// add message observer to get the response.
+		ObserveMessageObserverImpl messageObserver = new ObserveMessageObserverImpl(handler, relation);
+		request.addMessageObserver(messageObserver);
+		// add notification listener to all notification
+		NotificationListener notificationListener = new Adapter(messageObserver, request);
+		outEndpoint.addNotificationListener(notificationListener);
+		// messageobserver should remove this listener when the request is cancelled
+		messageObserver.setNotificationListener(notificationListener);
+		relation.setNotificationListener(notificationListener);
```

Comment:
if you push all the notif to the MessageObserver you encourage user to write code not working in cluster mode. For example a device push some notif and then push the next one 2hours after. It will probably reach another node and if will be only available using the NotificationListener.

If you ask the user to directly use the NotificationListener for receiving notification he will write clusterisable code from day one.